text/0039-dictionary.md

```diff
@@ -0,0 +1,118 @@
+
+This proposal intends to fill in the functionality gap and resolve past problems of local keys under a new name: a dictionary API.
+
+## Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+Implementation of a dictionary will involve adding a separate keyspace to the storage layer. Each operation on the dictionary key will require a "seed" URef, which the execution engine will authenticate.
+
+Under the hood, dictionary addresses have a length limit of 32 bytes, and the system will generate an address in the storage by hashing the URef address and key bytes together through API access.
+
+Example usage from the perspective of a contract developer:
+
+```rust
+fn call() {
+    // Initialize and store the URef in named keys
+    let dictionary_uref: URef = storage::new_dictionary("data");
+
+    // Write data
+    storage::dictionary_put(dictionary_uref, b"key", "value".to_string());
+
+    // Read data
+    let value: String = storage::dictionary_get(dictionary_uref, b"key");
+    assert_eq!(value, "value".to_string());
+}
+```
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+No existing system parts will change due to the proposal's nature, which intends only to extend the system's affected parts.
+
+This document intends to introduce new host functions to create and modify dictionary keys. Implementation will also match the new functionality in contract API for developers.
+
+This proposal introduces a hashing function under the hood due to the current trie implementation limitation that does not support keys longer than 32 bytes. As hashing is a lossy function, we still want to retain developers' ability to list data referenced through dictionary URef. The system will transparently wrap written data in a structure which will preserve the original `URef` and actual key bytes. This operation is completely hidden from the user and considered an implementation detail.
+
+Whenever we are ready to migrate the data from short 32-byte space into 64-byte, we will employ an expensive one-time sweep over old Dictionary keys and rehash them into new, more extended space.
+
+### Key::Dictionary
+
+Adds a new variant that holds 32 bytes hashed value based on:
+
+```
+blake2b256(uref_addr || key_bytes)
+```
+
+Further operations such as reading or writing use the earlier algorithm to access the `Key::Dictionary` space.
+
+### dictionary_new
+
+```rust
+fn dictionary_new(name: &str) -> URef;
+```
+
+Host provisions an URef of "unit" type under the hood used internally to secure access to the dictionary. Resulting URef will be automatically put into named keys, retaining his ability to add new data.
```

Comment:
What happens if `name` is empty? Or if there already exists an entry with the same name in named keys collection of the account?